python/ray/data/tests/test_dataset.py

```diff
@@ -2492,6 +2492,33 @@ def test_map_batches_actors_preserves_order(ray_start_regular_shared):
     assert ds.map_batches(lambda x: x, compute="actors").take() == list(range(10))
 
 
+@pytest.mark.parametrize(
+    "num_rows,num_blocks,batch_size",
+    [
+        (10, 5, 2),
+        (10, 1, 10),
+        (10, 1, None),
+        (12, 3, 2),
+    ],
+)
+def test_map_batches_batch_mutation(
+    ray_start_regular_shared, num_rows, num_blocks, batch_size
+):
+    # Test that batch mutation works without encountering a read-only error (e.g. if the
+    # batch is a zero-copy view on data in the object store).
+    def mutate(df):
+        df["value"] += 1
+        return df
+
+    ds = ray.data.range_table(num_rows, parallelism=num_blocks).repartition(num_blocks)
+    # Convert to Pandas blocks.
+    ds = ds.map_batches(lambda df: df, batch_format="pandas", batch_size=None)
+
+    # Apply UDF that mutates the batches.
+    ds = ds.map_batches(mutate, batch_size=batch_size)
+    assert [row["value"] for row in ds.iter_rows()] == list(range(1, num_rows + 1))
```

Comment:
Hmm if I'm understanding this correctly, does this test pass even without the changes in this PR?

It just checks if the udf is applied which will work regardless of if the batch size is actually respected or not?

Reply:
It passed before this PR due to the slice copying here: https://github.com/ray-project/ray/blob/fe6b5ce0e7a6cf6b2957c59e79ffc01d5e2dd9a1/python/ray/data/dataset.py#L478

When making the change to use the `Batcher` ([here](https://github.com/ray-project/ray/blob/7b5fccdb39d4a04a97e28fd9f7281fa689a99148/python/ray/data/dataset.py#L470-L503)), this was changed to be zero-copy, and some of the AIR examples that mutate the batch inputs failed, so I added this test to cover this "always copy the batch" guarantee.

Reply:
but should the udf be something like this

```
def mutate(df):
        assert len(df) == batch_size
```

to actually test if coalesce is working?

Reply:
That's not what this test is checking, it's checking that the existing "batch is mutable" guarantee is satisfied, since that's currently relied on but before this PR wasn't covered by tests.

I can add a test that better covers the changed semantics in this PR, but I should also emphasize that `len(df) == batch_size` is not guaranteed to be true by this PR; this is a best-effort coalescing of adjacent blocks that are cumulatively smaller than the provided batch size.

Reply:
So for blocks of size [2, 2, 3, 2] and `batch_size=5`:
- Before this PR, you'd get 4 tasks with batches of size [2, 2, 3, 2].
- After this PR, you'd get 2 tasks with batches of size [4, 5].

I can add a test covering that best-effort coalescing, I'll do that now.